tsdb/index/tsi1/series_file.go

```diff
@@ -450,3 +497,173 @@ func (a seriesKeys) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 func (a seriesKeys) Less(i, j int) bool {
 	return CompareSeriesKeys(a[i], a[j]) == -1
 }
+
+const (
+	SeriesMapFileSuffix = "map"
+
+	SeriesMapLoadFactor = 90
+
+	SeriesMapCountSize     = 8
+	SeriesMapMaxOffsetSize = 8
+	SeriesMapHeaderSize    = SeriesMapCountSize + SeriesMapMaxOffsetSize
+
+	SeriesMapElemSize = 8 + 8 // hash + value
+)
+
+// seriesMap represents a read-only hash map of series offsets.
+type seriesMap struct {
+	path  string
+	sfile *SeriesFile
+	inmem *rhh.HashMap
+
+	n         int64
+	maxOffset uint64
+	capacity  int64
+	data      []byte
+	mask      int64
+}
+
+func newSeriesMap(path string, sfile *SeriesFile) *seriesMap {
+	return &seriesMap{path: path, sfile: sfile}
+}
+
+func (m *seriesMap) open() error {
+	// Memory map file data.
+	data, err := mmap.Map(m.path, 0)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	m.data = data
+
+	// Read header if available.
+	if len(m.data) > 0 {
+		buf := data
+		m.n, buf = int64(binary.LittleEndian.Uint64(buf)), buf[SeriesMapCountSize:]
+		m.maxOffset, buf = uint64(binary.LittleEndian.Uint64(buf)), buf[SeriesMapMaxOffsetSize:]
+		m.capacity = int64(len(buf) / SeriesMapElemSize)
+		m.mask = int64(m.capacity - 1)
+	} else {
+		m.n, m.maxOffset = 0, 1
+	}
+
+	// Index all data created after the on-disk hash map.
+	inmem := rhh.NewHashMap(rhh.DefaultOptions)
+	for b, offset := m.sfile.data[m.maxOffset:m.sfile.size], m.maxOffset; len(b) > 0; {
+		var key []byte
+		key, b = ReadSeriesKey(b)
+		inmem.Put(key, offset)
+		offset += uint64(len(key))
+	}
+	m.inmem = inmem
+
+	return nil
+}
+
+func (m *seriesMap) close() error {
+	if m.data != nil {
+		if err := mmap.Unmap(m.data); err != nil {
+			return err
+		}
+		m.data = nil
+	}
+	return nil
+}
+
+// offset finds the series key's offset in either the on-disk or in-memory hash maps.
+func (m *seriesMap) offset(key []byte) uint64 {
+	if offset := m.onDiskOffset(key); offset != 0 {
+		return offset
+	}
+	offset, _ := m.inmem.Get(key).(uint64)
+	return offset
+}
+
+func (m *seriesMap) onDiskOffset(key []byte) uint64 {
+	if len(m.data) == 0 {
+		return 0
+	}
+
+	hash := rhh.HashKey(key)
+	for d, pos := int64(0), hash&m.mask; ; d, pos = d+1, (pos+1)&m.mask {
+		elem := m.data[SeriesMapHeaderSize+(pos*SeriesMapElemSize):]
+		elem = elem[:SeriesMapElemSize]
+
+		h := int64(binary.LittleEndian.Uint64(elem[:8]))
```

Comment:
We don't use `LittleEndian` representations anywhere else in the codebase; is there a specific reason for choosing it in this case?

Reply:
Fixed.